src/fs/cluster_allocator.hh

```diff
@@ -22,29 +22,49 @@
 #pragma once
 
 #include "fs/cluster.hh"
+#include "seastar/core/circular_buffer.hh"
+#include "seastar/core/future.hh"
+#include "seastar/core/semaphore.hh"
 
-#include <deque>
 #include <optional>
 #include <unordered_set>
 
-namespace seastar {
-
-namespace fs {
+namespace seastar::fs {
 
 class cluster_allocator {
-    std::unordered_set<cluster_id_t> _allocated_clusters;
-    std::deque<cluster_id_t> _free_clusters;
+    std::unordered_map<cluster_id_t, bool> _allocated_clusters;
+    circular_buffer<cluster_id_t> _free_clusters;
+    semaphore _cluster_sem;
+
+    cluster_id_t do_alloc() noexcept;
+    void do_free(cluster_id_t cluster_id) noexcept;
 
 public:
-    explicit cluster_allocator(std::unordered_set<cluster_id_t> allocated_clusters, std::deque<cluster_id_t> free_clusters);
+    cluster_allocator(std::unordered_set<cluster_id_t> allocated_clusters = {},
+            circular_buffer<cluster_id_t> free_clusters = {});
```

Comment:
it is not OK, that we have here one argument constructor that is also viable for implicit conversions... Please add default constructor and remove default arguments here.